oap-server/server-query-plugin/query-graphql-plugin/src/main/java/org/apache/skywalking/oap/query/graphql/resolver/MetricQuery.java

```diff
@@ -59,17 +63,24 @@ public IntValues getValues(final BatchMetricConditions metrics, final Duration d
             kv.setValue(query.readMetricsValue(condition, duration));
             values.addKVInt(kv);
         } else {
-            for (final String id : metrics.getIds()) {
-                KVInt kv = new KVInt();
-                kv.setId(id);
+            List<CompletableFuture<KVInt>> futureList = metrics.getIds().stream().map(id -> {
 
                 MetricsCondition condition = new MetricsCondition();
                 condition.setName(metrics.getName());
                 condition.setEntity(new MockEntity(id));
-
-                kv.setValue(query.readMetricsValue(condition, duration));
-                values.addKVInt(kv);
-            }
+                CompletableFuture<KVInt> future = CompletableFuture.supplyAsync(() -> {
+                    try {
+                        KVInt kv = new KVInt();
+                        kv.setId(id);
+                        kv.setValue(query.readMetricsValue(condition, duration));
+                        return kv;
+                    } catch (IOException e) {
+                        throw new RuntimeException(e);
```

Comment:
Once you do this, I think you need to check `CompletableFuture#isCompletedExceptionally` and process in this case?

Reply:
The exception will propagate to caller in `.join`

Reply:
OK, then.